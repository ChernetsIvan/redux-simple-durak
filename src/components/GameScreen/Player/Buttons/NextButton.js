import React from "react";
import PropTypes from "prop-types";

export const NextButton = ({endInd, cards, onNextClick}) => {
  let out = null;
  if (endInd < cards.length - 1 && cards.length > 10) {
    out = (
      <button onClick={onNextClick} className="btn btn-info">
        <b>{cards.length - 1 - endInd}&rarr;</b>
      </button>
    );
  }
  return out;
};

NextButton.propTypes = {
  endInd: PropTypes.number.isRequired,
  cards: PropTypes.array.isRequired,
  onNextClick: PropTypes.func.isRequired
};
