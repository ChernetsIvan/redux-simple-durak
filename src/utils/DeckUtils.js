import { pull, find } from 'lodash-es';

import {cardRanks, cardSuits} from './../constants/GameConstants';

import CardModel from './../models/CardModel';

class DeckUtils{
    //отсортировать входной массив карт по возрастанию isAsc=true
    //(убыванию-> isAsc=false)  по параметру 'power'
    static sortInputDeckByPower(inputDeck, isAsc){
        inputDeck.sort(function(a,b){
            if(isAsc){
                return a.power - b.power;
            }else{
                return b.power - a.power;
            }        
        });
    }

    //Возвращает наполненную всеми 36-ю картами колоду
    static fillCards(trumpSuit){
        let fullDeck = [];
        //Устанавливает power для каждой карты по принципу:
        //козырь->10*rank.cardValue, НЕкозырь->1*rank.cardValue
        let z = 0;
        let power, coef;               
        cardSuits.forEach(function(element,index, array){
            if(element.suit === trumpSuit.suit){
                coef = 10;
            }else{
                coef = 1;
            }
            cardRanks.forEach(function(element2, index2, array2){
                power = coef * element2.cardValue;
                fullDeck.push(
                    new CardModel(z, element2, element, power)
                );
                z++;
            });
        });
        return fullDeck;
    }

    //Возвращает перемешанную колоду
    static shuffle(inputDeck){
        var newFullDeck = [];
        while(inputDeck.length > 0){
            //Случайный индекс в inputDeck
            let randIndex = Math.random() * (inputDeck.length-1);
            randIndex = Math.round(randIndex);

            let card = inputDeck[randIndex];
            newFullDeck.push(card);

            inputDeck.splice(randIndex, 1);
        }
        return newFullDeck;
    }

    //Перемещает первую попавшуюся карту с козырной мастью
    //в конец колоды 
    static moveAnyCardWithTrumpSuitToTailOfFullDeck(trumpSuit, fullDeck){
        let card = find(fullDeck, (card) => {
            return card.suit.suit === trumpSuit.suit;
        });
        pull(fullDeck, card);
        fullDeck.push(card);
    }

    //взять из НАЧАЛА (!) колоды fullDeck и положить в inputDeck столько карт, 
    //чтобы в inputDeck в итоге было 6 карт
    static giveUpToSixCards(fullDeck, inputDeck){
        //если хотя бы одно из условий не выполняется, в цикл не заходим:
        while(inputDeck.length < 6 && fullDeck.length > 0){
            let card = fullDeck.shift();
            inputDeck.push(card);
        }
    }

    //Возвращает козырную масть (не карту!) для будущей игры
    static chooseSuitForTrumpCardInFuture(){
        let randIndex = Math.random() * (cardSuits.length-1);
        randIndex = Math.round(randIndex);
        return cardSuits[randIndex];
    }
}

export default DeckUtils;