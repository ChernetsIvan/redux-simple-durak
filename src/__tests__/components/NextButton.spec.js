import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-15";
import toJson from "enzyme-to-json";

import DeckUtils from "./../../utils/DeckUtils";
import NextButton from "./../../components/GameScreen/Player/Buttons/NextButton";

Enzyme.configure({ adapter: new Adapter() });

describe("NextButton", () => {
  const cards = DeckUtils.fillCards({ suit: "Ðš" });
  const mockCallBack = jest.fn();

  test("Test click event", () => {
    const button = shallow(
      <NextButton endInd={1} cards={cards} onNextClick={mockCallBack} />
    );
    button.find("button").simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  test("test of rendering with > 10 cards", () => {
    const button = shallow(
      <NextButton endInd={1} cards={cards} onNextClick={mockCallBack} />
    );
    expect(toJson(button)).toMatchSnapshot();
  });

  test("test of rendering with < 10 cards", () => {
    const button = shallow(
      <NextButton
        endInd={0}
        cards={cards.slice(0, 5)}
        onNextClick={mockCallBack}
      />
    );
    expect(toJson(button)).toMatchSnapshot();
  });
});
