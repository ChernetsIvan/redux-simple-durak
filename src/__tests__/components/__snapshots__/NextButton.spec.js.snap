// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextButton test of rendering with < 10 cards 1`] = `""`;

exports[`NextButton test of rendering with > 10 cards 1`] = `
<button
  className="btn btn-info"
  onClick={[Function]}
>
  <b>
    34
    â†’
  </b>
</button>
`;
